client/blb/client.go

```diff
@@ -1130,136 +1128,113 @@ func (cli *Client) readOneTractReplicated(
-	if err == core.ErrVersionMismatch {
-		badVersionHost = host
-	}
-	if err != core.NoError && err != core.ErrEOF {
-		log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-		// If we failed to read from a TS, report that to the curator. Defer so we can examine
-		// result.err at the end, to see if we succeeded on another or not.
-		defer func(host string, err core.Error) {
-			couldRecover := result.err == core.NoError || result.err == core.ErrEOF
-			go cli.curators.ReportBadTS(context.Background(), curAddr, tract.Tract, host, "read", err, couldRecover)
-		}(host, err)
-		return true, badVersionHost, err // try another host
-
-	}
-	log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-	// In case of short read, i.e., read < len(thisB), we should
-	// pad the result with 0's. This doesn't need to be done
-	// explicitly if 'thisB' given to us starts to be empty.
-	// However, just in case it's not, we will ensure that by
-	// overwritting untouched bytes with 0's.
-	for i := read; i < len(thisB); i++ {
-		thisB[i] = 0
-	}
-	*result = tractResult{len(thisB), read, err, badVersionHost}
-	return false, badVersionHost, err
-}
-
-func (cli *Client) readOneTractWithBackupReplicated(
-	ctx context.Context,
-	curAddr string,
-	result *tractResult,
-	tract *core.TractInfo,
-	thisB []byte,
-	thisOffset int64) {
-
-	var badVersionHost string
+	if cli.backupReadState.BackupReadBehavior.Enabled {
+		ch := make(chan tractResult)
 
-	order := rand.Perm(len(tract.Hosts))
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
 
-	// Generate a request ID to use for cancelling alternate requests.
-	reqID := core.GenRequestID()
+		maxNumBackups := cli.backupReadState.BackupReadBehavior.MaxNumBackups
+		delay := cli.backupReadState.BackupReadBehavior.Delay
 
-	ch := make(chan *tractResult)
-	errc := make(chan core.Error)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	err := core.ErrAllocHost // default error if none present
-	for i, n := range order {
-		go func(i, j int) {
-			select {
-			case <-ctx.Done():
-				errc <- core.ErrCanceled
+		readFunc := func(n int) {
+			host := tract.Hosts[n]
+			if host == "" {
+				log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[n])
+				ch <- tractResult{0, 0, err, badVersionHost}
 				return
-			case <-cli.backupDelay(time.Duration(2*i) * time.Millisecond):
-				// sleep for the backup delay to prevent tied backup requests cancelling,
-				// each other's work.
 			}
 
-			host := tract.Hosts[j]
-			if host == "" {
-				log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[j])
+			// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
+			read, err := cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisB, thisOffset)
+			if err == core.ErrVersionMismatch {
+				badVersionHost = host
```

Comment:
We decided to remove the bad ts reporting for now because it makes things complicated, and add it back in a better form later.

I think you can actually remove all this badVersionHost stuff too, since it's just used for that. There's a chance we'd have to add it back later, but actually I think the eventual shape of bad ts reporting might not look like that.